tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/RecordingZ3SolverContext.scala

```diff
@@ -0,0 +1,207 @@
+    *
+    * @param n pop n times, if n > 0, otherwise, do nothing
+    */
+  override def pop(n: Int): Unit = {
+    solver.pop(n)
+    logStackRev = logStackRev.drop(n - 1) // n - 1 in logStack + 1 in lastLog
+    lastLogRev = logStackRev.head
+    logStackRev = logStackRev.tail
+  }
+
+  /**
+    * Clean the context
+    */
+  override def dispose(): Unit = {
+    solver.dispose()
+  }
+
+  /**
+    * Declare a constant for an arena cell.
+    * This method is called automatically by the arena.
+    *
+    * @param cell a (previously undeclared) cell
+    */
+  override def declareCell(cell: ArenaCell): Unit = {
+    lastLogRev = DeclareCellRecord(cell) :: lastLogRev
+    solver.declareCell(cell)
+  }
+
+  /**
+    * Declare an arena edge of type 'has'. This method introduces a Boolean variable for the edge.
+    * This method is called automatically by the arena. If the context already contains the constant
+    * with the same name, then this method does nothing.
+    *
+    * @param set  the containing set
+    * @param elem a set element
+    */
+  override def declareInPredIfNeeded(set: ArenaCell, elem: ArenaCell): Unit = {
+    lastLogRev = DeclareInPredRecord(set, elem) :: lastLogRev
+    solver.declareInPredIfNeeded(set, elem)
+  }
+
+  /**
+    * Check whether the current view of the SMT solver is consistent with arena.
+    *
+    * @param arena an arena
+    */
+  override def checkConsistency(arena: Arena): Unit = {
+    solver.checkConsistency(arena)
+  }
+
+  /**
+    * Assert that a Boolean TLA+ expression holds true.
+    *
+    * @param ex a simplified TLA+ expression over cells
+    */
+  override def assertGroundExpr(ex: TlaEx): Unit = {
+    lastLogRev = AssertGroundExprRecord(ex) :: lastLogRev
+    solver.assertGroundExpr(ex)
+  }
+
+  /**
+    * Evaluate a ground TLA+ expression in the current model, which is available after a call to sat().
+    * This method assumes that the outcome is either a Boolean or integer.
+    * If not, it throws SmtEncodingException.
+    *
+    * @param ex an expression to evaluate
+    * @return a TLA+ value
+    */
+  override def evalGroundExpr(ex: TlaEx): TlaEx = {
+    solver.evalGroundExpr(ex)
+  }
+
+  /**
+    * Write a message to the log file. This is helpful to debug the SMT encoding.
+    *
+    * @param message message text, call-by-name
+    */
+  override def log(message: => String): Unit = {
+    if (config.debug) {
+      solver.log(message)
+    }
+  }
+
+  /**
+    * Is the current context satisfiable?
+    *
+    * @return true if and only if the context is satisfiable
+    */
+  override def sat(): Boolean = {
+    solver.sat()
+  }
+
+  /**
+    * Check satisfiability of the context with a timeout
+    *
+    * @param timeoutSec the timeout in seconds. If timeout <= 0, it is not effective
+    * @return Some(result), if no timeout happened; otherwise, None
+    */
+  override def satOrTimeout(timeoutSec: Long): Option[Boolean] = {
+    solver.satOrTimeout(timeoutSec)
```

Comment:
Too bad all these passthrough's are needed and we can't just inherit the methods. Makes for a lot of boilerplate!

Reply:
True. It is weird. However, this wrapper should work for any `SolverContext`, not just `Z3SolverContext`. I should refactor it to either directly inherit from Z3SolverContext, or wrap an arbitrary `SolverContext`. I would prefer the latter.

Reply:
I've renamed `RecordingZ3SolverContext` to `RecordingSolverContext` and introduced the factory method `RecordingSolveContext.createZ3`.